packages/template-retail-react-app/app/utils/url.js

```diff
@@ -149,23 +150,95 @@ export const getPathWithLocale = (shortCode, opts = {}) => {
-            locale: isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef ? '' : shortCode
-        },
-        opts
+    const newUrl = fillUrlTemplate(
+        `${pathname}${Array.from(queryString).length !== 0 ? `?${queryString}` : ''}`,
+        // By default, as for home page, when the values of site and locale belongs to the default site,
+        // they will be not shown in the url just
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef
+            ? ''
+            : siteRef || defaultSite.alias || defaultSite.id,
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef ? '' : shortCode
     )
     return newUrl
 }
 
+/**
+ * Generates the URL Template literal (Template string) used to build URLs in the App according
+ * the current selected site/locale and the default App URL configuration.
+ *
+ * @param appConfig Application default configuration.
+ * @param siteRef Current selected Site reference. The value can be the Site id or alias.
+ * @param localeRef Current selected Locale reference. The value can be the Locale id or alias.
+ * @returns {function(*, *, *, *=): string} function providing path, site and locale generates a site URL.
+ */
+export const createUrlTemplate = (appConfig, siteRef, localeRef) => {
+    const {site: siteConfig, locale: localeConfig, showDefaults: showDefaultsConfig} = appConfig.url
+    const defaultSite = getDefaultSite()
+    const sites = getSites()
+    const site =
+        sites.find((site) => {
+            return site.alias === siteRef || site.id === siteRef
+        }) || defaultSite
+    const defaultLocale = getLocaleByReference(site, site.l10n.defaultLocale)
+
+    const isDefaultSite =
+        defaultSite.id === siteRef || (defaultSite.alias && defaultSite.alias === siteRef)
+    const isDefaultLocale =
+        defaultLocale.id === localeRef || (defaultLocale.alias && defaultLocale.alias === localeRef)
+
+    const querySite =
+        (siteConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultSite)
+    const queryLocale =
+        (localeConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultLocale)
+
+    const isQuery = querySite || queryLocale
+
+    const pathSite =
+        (siteConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultSite)
+    const pathLocale =
+        (localeConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultLocale)
+
+    const templateConfig = {
+        pathSite,
+        pathLocale,
+        querySite,
+        queryLocale,
+        isQuery
+    }
+
+    return (path, site, locale, config = templateConfig) => {
```

Comment:
Is you intention that you can pass a different config to you template filler function as an escape hatch?

Reply:
No. I removed the config since it's not needed.